Added more details about deconstruct (#27017)
dotnet/docs: docs/csharp/fundamentals/functional/deconstruct.md

:::code source="./snippets/deconstructing-tuples/deconstruct-extension1.cs":::

## Extension method for system types

Some system types provide the `Deconstruct` method as a convenience. For example, the <xref:System.Collections.Generic.KeyValuePair%602?displayProperty=nameWithType> type provides this functionality. When you're iterating over a <xref:System.Collections.Generic.Dictionary%602?displayProperty=nameWithType> each element is a `KeyValuePair<TKey, TValue>` and can be deconstructed. Consider the following example:

:::code source="./snippets/deconstructing-tuples/deconstruct-kvp.cs" id="KeyValuePair":::

> Not all type provide a `Deconstruct` methods as that's not practical, in fact very few types do. You may wish to add the `Deconstruct` method to various system types if deemed valuable. Consider the following extension method:

Reviewer: ```suggestion
You can add a `Deconstruct` method to system types that don't have one. Consider the following extension method:
```

:::code source="./snippets/deconstructing-tuples/deconstruct-extension2.cs" id="NullableExtensions":::

> This extension method allows all <xref:System.Nullable%601> types to be deconstructed into a tuple of `(bool hasValue, T value)`. Consider the following example:

Reviewer: ```suggestion
This extension method allows all <xref:System.Nullable%601> types to be deconstructed into a tuple of `(bool hasValue, T value)`. The following example shows code that uses this extension method:
```
